src/mono/wasm/runtime/binding_support.js

```diff
@@ -2108,19 +2332,236 @@ var BindingSupportLib = {
 		} finally {
 			nameRoot.release();
 		}
 	},
+	mono_wasm_cancel_promise: function (thenable_js_handle, is_exception) {
+		try {
+			const promise = BINDING.mono_wasm_get_jsobj_from_js_handle(thenable_js_handle)
+			const promise_control = promise[BINDING.promise_control_symbol];
+			promise_control.reject("OperationCanceledException");
+		}
+		catch (ex) {
+			return BINDING.wrap_error(is_exception, ex);
+		}
+	},
+	mono_wasm_web_socket_open: function (uri, subProtocols, on_close, web_socket_js_handle, thenable_js_handle, is_exception) {
+		var uri_root = MONO.mono_wasm_new_root(uri);
+		let sub_root = MONO.mono_wasm_new_root(subProtocols);
+		let on_close_root = MONO.mono_wasm_new_root(on_close);
+		try {
+			const js_uri = BINDING.conv_string(uri_root.value);
+			if (!js_uri) {
+				return BINDING.wrap_error(is_exception, "ERR12: Invalid uri '" + uri_root.value + "'");
+			}
+
+			const js_subs = BINDING._mono_array_root_to_js_array(sub_root);
+
+			const js_on_close = BINDING._wrap_delegate_root_as_function(on_close_root);
+
+			const ws = new globalThis.WebSocket(js_uri, js_subs);
+			var { promise, promise_control } = BINDING._create_cancelable_promise();
+
+			ws[BINDING.wasm_ws_pending_receive_event_queue] = new BINDING.Queue();
+			ws[BINDING.wasm_ws_pending_receive_promise_queue] = new BINDING.Queue();
+			ws[BINDING.wasm_ws_pending_open_promise] = promise_control;
+			ws[BINDING.wasm_ws_pending_send_promises] = [];
+			ws[BINDING.wasm_ws_pending_close_promises] = [];
+			ws.binaryType = "arraybuffer";
+			const on_message = (ev) => {
+				if (ws[BINDING.wasm_ws_is_aborted]) return;
+				BINDING._mono_wasm_web_socket_on_message(ws, ev);
+				MONO.prevent_timer_throttling();
+			}
+			ws.addEventListener('message', on_message);
+			ws.addEventListener('open', () => {
+				if (ws[BINDING.wasm_ws_is_aborted]) return;
+				promise_control.resolve(null);
+				MONO.prevent_timer_throttling();
+			}, { once: true })
+			ws.addEventListener('close', (ev) => {
+				ws.removeEventListener('message', on_message);
+				if (ws[BINDING.wasm_ws_is_aborted]) return;
+				js_on_close(ev.code, ev.reason);
+
+				// this reject would not do anything if there was already "open" before it.
+				promise_control.reject(ev.reason);
+
+				for(close_promise_control of ws[BINDING.wasm_ws_pending_close_promises]){
+					close_promise_control.resolve();
+				}
+
+				// send close to any pending receivers, to wake them
+				ws[BINDING.wasm_ws_pending_receive_promise_queue].drain(receive_promise_control => receive_promise_control.resolve({
+					type: 2,// close
+					count: 0,
+					end_of_message: true
+				}));
+			}, { once: true })
+
+			var { task_ptr, then_js_handle } = BINDING._wrap_js_thenable_as_task(promise);
+			// task_ptr above is not rooted, we need to return it to mono without any intermediate mono call which could cause GC
+			setValue(thenable_js_handle, then_js_handle, "i32");
+			var ws_js_handle = BINDING.mono_wasm_get_js_handle(ws);
```

Comment:
I think you should move this up above _wrap_js_thenable_as_task, so that task_ptr is basically returned without us doing anything that could cause it to get GC'd